infix/lexer.go

```diff
@@ -0,0 +1,225 @@
+package infix
+
+import (
+	"fmt"
+	"strings"
+	"unicode"
+	"unicode/utf8"
+)
+
+type (
+	lexer struct {
+		input  string // expression being lex'ed
+		start  int    // start position of token
+		pos    int    // pos in the input
+		width  int    // width of last rune
+		tokens chan Tokval
+	}
+
+	stateFn func(*lexer) stateFn
+
+	// Tokval is the token type + value
+	Tokval struct {
+		T Token
+		V string
+	}
+)
+
+// eof mimics the EOF but for the input string.
+// The name isn't eoi to avoid confusion.
+const eof = -1
+
+func (t Tokval) String() string {
+	return fmt.Sprintf("Token(%s, %s)", t.T, t.V)
+}
+
+// Lex creates a concurrent lexer and returns a
+// channel of tokens processed from input.
+func Lex(input string) chan Tokval {
```

Comment:
What you need to avoid?

Reply:
The caller writing things or closing the channel, since both things seems to go against the original intent of your API. I like APIs to have the smallest surface possible and to be explicit. Leaving as it is right now you are saying to me (with code) that I can write and close this channel (you can use docs to say that I should not, but if you can say it with code I prefer the code to do the talking =P). But this may be just me...being pedantic and annoying =D